### PR TITLE
Require colocation of snapshots in a release

### LINKED ISSUE #3448: Snapshot location might differ from that of BQ slots
Currently TDR snapshots in `dev` and `prod` reside in `US`. The snapshots created in response to HumanCellAtlas/dcp2#17 all reside in `us-central1`. We need to make TDR source location configurable, enforce that all sources reside in the configured location, and purchase slots in the configured location. Also see sketch below (note the caching change).

```diff
Index: scripts/register_sam.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/scripts/register_sam.py b/scripts/register_sam.py
--- a/scripts/register_sam.py	(revision 66112250e45fbbdbb7e433656cb998e3693352e0)
+++ b/scripts/register_sam.py	(date 1632256578847)
@@ -33,11 +33,14 @@
     )
     for source in set(chain(*map(config.tdr_sources, tdr_catalogs))):
         source = TDRSourceSpec.parse(source)
-        api_project = tdr.lookup_source_project(source)
-        require(api_project == source.project,
+        tdr_source = tdr.lookup_source(source)
+        # FIXME: This has nothing to do with SAM
+        require(tdr_source.project == source.project,
                 'Actual Google project of TDR source differs from configured '
                 'one',
-                api_project, source)
+                tdr_source.project, source)
+        require(tdr_source.location == config.tdr_source_location)
+        # TODO: This is now cached, see if that makes a difference
         tdr.check_api_access(source)
         tdr.check_bigquery_access(source)
 
Index: src/azul/terra.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/terra.py b/src/azul/terra.py
--- a/src/azul/terra.py	(revision 66112250e45fbbdbb7e433656cb998e3693352e0)
+++ b/src/azul/terra.py	(date 1632256578852)
@@ -386,22 +386,18 @@
     """
     A client for the Broad Institute's Terra Data Repository aka "Jade".
     """
+    @attr.s(frozen=True, kw_only=True, auto_attribs=True)
+    class TDRSource:
+        project: str
+        id: str
+        location: str
 
-    @cache
-    def lookup_source_project(self, source: TDRSourceSpec) -> str:
-        """
-        Return the name of the Google Cloud project containing the source
-        (snapshot or dataset) with the specified name.
-        """
-        return self._lookup_source(source)['dataProject']
-
-    @cache
-    def lookup_source_id(self, source: TDRSourceSpec) -> str:
-        """
-        Return the primary identifier of the source (snapshot or dataset) with
-        the specified name.
-        """
-        return self._lookup_source(source)['id']
+    def lookup_source(self, source: TDRSourceSpec) -> TDRSource:
+        source = self._lookup_source(source)
+        storage = one(s for s in source['storage'] if s['cloudResource'] == 'bigquery')
+        return self.TDRSource(project=source['dataProject'],
+                              id=source['id'],
+                              location=storage['region'])
 
     def check_api_access(self, source: TDRSourceSpec) -> None:
         """
@@ -410,6 +406,7 @@
         self._lookup_source(source)
         log.info('TDR client is authorized for API access to %s.', source)
 
+    @cache
     def _lookup_source(self, source: TDRSourceSpec) -> JSON:
         resource = f'{source.type_name} {source.name!r} via the TDR API'
         tdr_path = source.type_name + 's'
```

### PR DESCRIPTION
**Pull request by hannes-ucsc**
DataBiosphere/azul#3448

### CODE REVIEW

**Review by aherbst-broad — APPROVED** [review]

**Review by NoopDog — APPROVED** [review]

File: docs/dcp2_system_design.rst
Comment (ruchim) [inline docs/dcp2_system_design.rst:1130]: what would be the use case for relocating snapshots?  It's not a feature that exists -- so trying to understand the driving forces

File: docs/dcp2_system_design.rst
Comment (ruchim) [inline docs/dcp2_system_design.rst:1131]: on the TDR front -- we pick location for data based on the data policy. So if we were hosting data that requires it to be stored in a European region int he cloud, then it breaks the model for all snapshots living in one location. We may need a discussion on this.

**Review by ruchim — CHANGES_REQUESTED** [review]
Left a review comment that needs discussion before I can approve.

**Review by kbergin — APPROVED** [review]
I approve on the condition that it's merged after Ruchi approves, which I know it will not be merged until then! No pipeline concerns I can think of here.

File: docs/dcp2_system_design.rst
Comment (hannes-ucsc) [reply docs/dcp2_system_design.rst:1130]: This is not saying that snapshots have to be relocated, just that they can. I don't think it is far fetched to consider the case. If snapshots never have to be relocated, great. If they do, this points out a not so obvious consequence of the normative section right above.

Comment (hannes-ucsc) [reply docs/dcp2_system_design.rst:1130]: One use case would be that a particular region offers better cost or CO2 footprint so we might want to move snapshots to that region. Keep in mind that query jobs run in the location of the tables.

File: docs/dcp2_system_design.rst
Comment (hannes-ucsc) [reply docs/dcp2_system_design.rst:1131]: Yes, if and when that happens then we need to discuss a change to the specification. My team needs this constraint to avoid sinking more time into our around for TDR's lack of a cost effective query by primary key option (clustering).

**Comment by hannes-ucsc** [comment]
No PR in this repo is merged against an objection by any reviewer (step 5 of the [SOP](https://github.com/HumanCellAtlas/dcp2/blob/main/docs/dcp2_operating_procedures.rst#16pull-request-reviews)) so, pedantically speaking, there is no need to make one reviewer's approval conditional upon that of another reviewer.

**Review by ESapenaVentura — APPROVED** [review]
On the EBI side, LGTM 👍

File: docs/dcp2_system_design.rst
Comment (ruchim) [reply docs/dcp2_system_design.rst:1130]: short answer: as long as we are saying they "can" and not that they should, I'm all good!  

Long answer: Every use case I've seen around snapshot location is data policy driven -- it needs to live in a specific countries boundary or something along those lines. I think cost is also an important driver, but that most likely means we use cold storage like mechanisms to save cost, but agreed that cost will be a use case that impacts how objects are stored in the cloud

File: docs/dcp2_system_design.rst
Comment (ruchim) [reply docs/dcp2_system_design.rst:1131]: perfect

**Review by ruchim — APPROVED** [review]
all comments resolved

**Pull request merged**
